slack-lint-checks/src/main/java/slack/lint/JsonInflaterMoshiCompatibilityDetector.kt

```diff
@@ -0,0 +1,160 @@
+// Copyright (C) 2025 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package slack.lint
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.android.tools.lint.detector.api.TextFormat
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiModifier
+import com.intellij.psi.PsiPrimitiveType
+import com.intellij.psi.PsiType
+import org.jetbrains.uast.UCallExpression
+import slack.lint.util.sourceImplementation
+
+/**
+ * A detector that checks if a type passed into the JsonInflater.inflate/deflate methods follows
+ * Moshi's requirements for serialization/deserialization.
+ */
+class JsonInflaterMoshiCompatibilityDetector : Detector(), SourceCodeScanner {
+
+  override fun getApplicableUastTypes() = listOf(UCallExpression::class.java)
+
+  override fun createUastHandler(context: JavaContext): UElementHandler {
+    return object : UElementHandler() {
+      override fun visitCallExpression(node: UCallExpression) {
+        if (isJsonInflaterInflateOrDeflate(node)) {
+          when (node.methodName) {
+            "inflate" -> validateInflateReturnType(context, node)
+            "deflate" -> validateDeflateArguments(context, node)
+          }
+        }
+      }
+    }
+  }
+
+  private fun isJsonInflaterInflateOrDeflate(node: UCallExpression): Boolean {
+    // Get the method being called
+    val method = node.resolve() ?: return false
+
+    // Check if it's from JsonInflater class
+    val containingClass = method.containingClass?.qualifiedName ?: return false
+    if (containingClass != FQCN_JSON_INFLATER) return false
```

Comment:
My memory may be wrapped up in compiler land but you may be able to check `dispatchReceiver` directly on the node rather than resolving the method and then the class, but no big deal if not. Might be a compiler plugin API.

You also want to check the extension receiver, if any. Remember there are `inline fun <reified T> JsonInflater.inflate()/deflate()` functions which wouldn't have a containing class